File: 33.Search in Rotated Sorted Array.java
public class Solution {
    public int search(int[] nums, int target) {
        if(nums==null || nums.length==0)
            return -1;
        int l = 0;
        int r = nums.length-1;
        while(l<r){
            int mid = (l+r)/2;
            if(nums[l]<nums[r]){
                if(target<nums[l] || target>nums[r]){
                    return -1;
                }
                if(nums[mid]<target){
                    l = mid+1;
                }else if(nums[mid]==target){
                    return mid;
                }else{
                    r = mid-1;
                }
            }else{
                if(target>=nums[l]){
                    if(nums[mid]<nums[r]){
                        r = mid-1;
                    }else{
                        if(nums[mid]<target){
                            l = mid+1;
                        }else if(nums[mid]==target){
                            return mid;
                        }else{
                            r = mid-1;
                        }
                    }
                }else if(target<=nums[r]){
                    if(nums[mid]<nums[r]){
                        if(nums[mid]<target){
                            l = mid+1;
                        }else if(nums[mid]==target){
                            return mid;
                        }else{
                            r = mid-1;
                        }
                    }else{
                        l  = mid+1;
                    }
                }else{
                    return -1;
                }
            }
        }
        if(nums[l]==target){
            return l;
        }else if(r>=0 && nums[r]==target){
            return r;
        }else{
            return -1;
        }
            
    }
}

File: 33.Search in Rotated Sorted Array.rb
# https://leetcode.com/problems/search-in-rotated-sorted-array/discuss/14435/Clever-idea-making-it-simple
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer}
def search(nums, target)
    lo, hi = 0, nums.size
    while lo < hi
        mid = (lo + hi) / 2
        if nums[0] <= target == nums[0] <= nums[mid]
            mid_num = nums[mid]
        else
            mid_num = nums[0] < nums[mid] ? -1.0/0.0 : 1.0/0.0
        end
        if target < mid_num
            hi = mid
        elsif target > mid_num
            lo = mid + 1
        else
            return mid
        end
    end
    -1
end
